sycl/plugins/opencl/pi_opencl.cpp

```diff
@@ -186,16 +187,66 @@ static cl_int checkDeviceExtensions(cl_device_id dev,
   return ret_err;
 }
 
+typedef CL_API_ENTRY cl_int(CL_API_CALL *clGetDeviceFunctionPointer_fn)(
+    cl_device_id device, cl_program program, const char *FuncName,
+    cl_ulong *ret_ptr);
+
+typedef CL_API_ENTRY cl_int(CL_API_CALL *clEnqueueWriteGlobalVariable_fn)(
+    cl_command_queue, cl_program, const char *, cl_bool, size_t, size_t,
+    const void *, cl_uint, const cl_event *, cl_event *);
+
+typedef CL_API_ENTRY cl_int(CL_API_CALL *clEnqueueReadGlobalVariable_fn)(
+    cl_command_queue, cl_program, const char *, cl_bool, size_t, size_t, void *,
+    cl_uint, const cl_event *, cl_event *);
+
+typedef CL_API_ENTRY cl_int(CL_API_CALL *clSetProgramSpecializationConstant_fn)(
+    cl_program program, cl_uint spec_id, size_t spec_size,
+    const void *spec_value);
+
+template <typename T> struct FuncPtrCache {
+  std::map<pi_context, T> Map;
+  std::mutex Mutex;
+};
+
+// FIXME There's currently no mechanism for cleaning up this cache, meaning that
+// it is invalidated whenever a context is destroyed. This could lead to reusing
+// an invalid function pointer if another context happends to have the same
+// native handle.
+struct ExtFuncPtrCacheT {
+  FuncPtrCache<clHostMemAllocINTEL_fn> clHostMemAllocINTELCache;
+  FuncPtrCache<clDeviceMemAllocINTEL_fn> clDeviceMemAllocINTELCache;
+  FuncPtrCache<clSharedMemAllocINTEL_fn> clSharedMemAllocINTELCache;
+  FuncPtrCache<clGetDeviceFunctionPointer_fn> clGetDeviceFunctionPointerCache;
+  FuncPtrCache<clCreateBufferWithPropertiesINTEL_fn>
+      clCreateBufferWithPropertiesINTELCache;
+  FuncPtrCache<clMemBlockingFreeINTEL_fn> clMemBlockingFreeINTELCache;
+  FuncPtrCache<clSetKernelArgMemPointerINTEL_fn>
+      clSetKernelArgMemPointerINTELCache;
+  FuncPtrCache<clEnqueueMemsetINTEL_fn> clEnqueueMemsetINTELCache;
+  FuncPtrCache<clEnqueueMemcpyINTEL_fn> clEnqueueMemcpyINTELCache;
+  FuncPtrCache<clGetMemAllocInfoINTEL_fn> clGetMemAllocInfoINTELCache;
+  FuncPtrCache<clEnqueueWriteGlobalVariable_fn>
+      clEnqueueWriteGlobalVariableCache;
+  FuncPtrCache<clEnqueueReadGlobalVariable_fn> clEnqueueReadGlobalVariableCache;
+  FuncPtrCache<clEnqueueReadHostPipeINTEL_fn> clEnqueueReadHostPipeINTELCache;
+  FuncPtrCache<clEnqueueWriteHostPipeINTEL_fn> clEnqueueWriteHostPipeINTELCache;
+  FuncPtrCache<clSetProgramSpecializationConstant_fn>
+      clSetProgramSpecializationConstantCache;
+};
+static ExtFuncPtrCacheT *const ExtFuncPtrCache = new ExtFuncPtrCacheT();
```

Comment:
Please add a comment about the issue you're solving to explain why raw pointer is used.